buildkite/bazel-central-registry/bcr_postsubmit.py

```diff
@@ -23,24 +23,114 @@
+        return
+    
+    for p in paths:
+        check_and_write_module_attestations(p)
+
+def get_new_attestations_json_paths():
+    commit = os.getenv("BUILDKITE_COMMIT")
+    cwd = os.getcwd()
+    paths = get_output(["git", "diff-tree", "--no-commit-id", "--name-only", commit, "-r"])
+    return [os.path.join(cwd, p) for p in paths.split("\n") if p.endswith(f"/{ATTESTATION_METADATA_FILE}")]
+
+def check_and_write_module_attestations(attestations_json_path):
+    print(f"Checking {attestations_json_path}...")
+    dest_dir = os.path.dirname(attestations_json_path)
+    with open(attestations_json_path, "rb") as af:
+        metadata = json.loads(af.read())
+    
+    for f in FILES_WITH_ATTESTATIONS:
+        try:
+            entry = metadata["attestations"][f]
+            check_and_write_single_attestation(entry["url"], entry["integrity"], dest_dir)
+        except Exception as ex:
+            raise AttestationError(f"{attestations_json_path} - {f}: {ex}") from ex
+
+    print("Done!")
+
+def check_and_write_single_attestation(url, integrity, dest_dir):
+    print(f"\tFound attestation @ {url}")
+    with requests.get(url) as response:
+        if response.status_code != 200:
+            raise AttestationError(f"{url}: HTTP {response.status_code}")
+
+        raw_content = response.content
+
+    check_integrity(raw_content, integrity)
+    print("\t\tIntegrity: OK")
+
+    dest = os.path.join(dest_dir, get_canonical_basename(url))
+    print(f"\t\tWriting attestation to {dest}...")
+    with open(dest, "wb") as f:
+        f.write(raw_content)
+
+def check_integrity(data, expected):
+    algorithm, _, _ = expected.partition("-")
+    assert algorithm in {"sha224", "sha256", "sha384", "sha512"}, "Unsupported SRI algorithm"
+
+    hash = getattr(hashlib, algorithm)(data)
+    encoded = base64.b64encode(hash.digest()).decode()
+    actual = f"{algorithm}-{encoded}"
+    if actual != expected:
+        raise AttestationError(f"Expected checksum {expected}, got {actual}.")
+
+# Attestation files in GitHub releases may have prefixes in their basename
+# to avoid conflicts when multiple modules are released together
+# (e.g. rules_python and rules_python_gazelle_plugin).
+# In this case we need to get the canonical basename.
+def get_canonical_basename(url):
+    actual_basename = os.path.basename(url)
+    for f in FILES_WITH_ATTESTATIONS:
+        if f in actual_basename:
+            return f"{f}.intoto.jsonl"
+    
+    raise AttestationError(f"Invalid basename of {url}.")
+
+
 def sync_bcr_content():
     print_expanded_group(":gcloud: Sync BCR content")
     subprocess.check_output(
         ["gsutil", "-h", "Cache-Control:no-cache", "cp", "./bazel_registry.json", BCR_BUCKET]
     )
     subprocess.check_output(
-        ["gsutil", "-h", "Cache-Control:no-cache", "-m", "rsync", "-d", "-r", "./modules", BCR_BUCKET + "modules"]
+        ["gsutil", "-h", "Cache-Control:no-cache", "-m", "rsync", "-r", "./modules", BCR_BUCKET + "modules"]
```

Comment:
Will this delete files in the GCS bucket if they don't exist in the local directory?

Reply:
My understanding is that it's the reverse - with -d we used to delete GCS-only files, which would delete attestations of other modules.

Reply:
Can we verify this? The documentation "-d" implies a different meaning, not sure if the gsutil version is different.

Reply:
Oh, I see, thanks for the clarification!